### ISSUE TITLE
SparseCategoricalCrossentropy and sparse_categorical_crossentropy are different due to precision loss

### ISSUE THREAD

**Issue by bjourne** [tensorflow/tensorflow#48129]
The following code demonstrates the issue:
```
from tensorflow.keras.losses import *
import numpy as np
import tensorflow as tf

loss_fn = SparseCategoricalCrossentropy(reduction='none')
y_true = np.random.randint(10, size = (4, 8))
y_pred = np.random.uniform(low=-10, high=10, size=(4,8,10))

l1 = loss_fn(y_true, y_pred).numpy()
l2 = sparse_categorical_crossentropy(y_true, y_pred).numpy()
print(np.array_equal(l1, l2), l1[0][0], l2[0][0])
```
According to TensorFlow documentation and expectations (see f.e https://stackoverflow.com/questions/59360263/tensorflow-2-0-what-is-the-difference-between-sparse-categorical-crossentropy-a) the two methods of computing the loss should always yield the exact same result. But that is not the case because the `SparseCategoricalCrossentropy` code path calls `compute_weighted_loss` and in that function the loss is converted to float32, losing precision. A comment in the function reads: "TODO(psv): Handle casting here in a better way, eg. if losses is float64 # we do not want to lose precision."

**Comment by Saduf2019** [comment]
@bjourne 
I ran the code shared on tf 2.4 and 2.5 stable version, here until five places after decimal the values are same.

**Comment by bjourne** [comment]
Yep, but *all* decimal places should be exactly the same. :)

**Comment by Saduf2019** [comment]
I ran the code on tf nightly and upto seven decimals the values match, please refer to the [gist here](https://colab.research.google.com/gist/Saduf2019/806ee04ea749c79669a33384b22e3e01/untitled591.ipynb)

**Comment by bjourne** [comment]
But the calculations should be exactly identical, not only to *n* decimal places. The comment in the code even says there is a bug here.

**Comment by d-01** [comment]
Probably related.

Two identical inputs gives different results with `categorical_crossentropy()` and `sparse_categorical_crossentropy()`:

```python
>>> tf.__version__
'2.5.0'

>>> tf.keras.losses.categorical_crossentropy([[1, 0]], [[.0, .1]]).numpy()
array([16.118095], dtype=float32)

>>> tf.keras.losses.sparse_categorical_crossentropy([0], [[.0, .1]]).numpy()
array([13.815511], dtype=float32)
```

First result is correct. Manual check:

```python
>>> -np.log(1e-7)  # epsilon, fuzz factor
16.11809565095832
```

Should I open new issue for this?

**Comment by sushreebarsa** [comment]
@bjourne I tried to replicate the issue on colab using tf-nightly(2.9.0.dev20220309) and didn't face the issue reported here.Please have a look at the gist [here](https://colab.research.google.com/gist/sushreebarsa/b24f8177f9d10339dbf4a12e97b9a6c6/gist48129.ipynb) and let us know if it helps?Thanks!

**Comment by bjourne** [comment]
> @bjourne I tried to replicate the issue on colab using tf-nightly(2.9.0.dev20220309) and didn't face the issue reported here.Please have a look at the gist [here](https://colab.research.google.com/gist/sushreebarsa/b24f8177f9d10339dbf4a12e97b9a6c6/gist48129.ipynb) and let us know if it helps?Thanks!

Looks like it has been fixed in tf 2.9.

**Comment by sushreebarsa** [comment]
@bjourne Thank you for the confirmation!
Glad it is working fine for you.
Thanks!
